packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts

```diff
@@ -222,27 +246,35 @@ export class DataVisualizerGraph
 
 	/**
 	 * Adds a visualizer node to the collection for the specified
-	 * {@link @fluidframework/shared-object-base#ISharedObject} if one does not already exist.
+	 * {@link VisualizableFluidObject} if one does not already exist.
 	 */
-	private registerVisualizerForSharedObject(sharedObject: ISharedObject): FluidObjectId {
-		if (!this.visualizerNodes.has(sharedObject.id)) {
+	private registerVisualizerForVisualizableObject(
+		visualObject: VisualizableFluidObject,
+	): FluidObjectId {
+		if (!this.visualizerNodes.has(visualObject.id)) {
 			// Create visualizer node for the shared object
-			const visualizationFunction =
-				this.visualizers[sharedObject.attributes.type] ?? visualizeUnknownSharedObject;
+			const visualizationFunction = isDataObject(visualObject)
+				? visualizeDataObject
+				: isPureDataObject(visualObject)
+					? visualizeUnknownSharedObject
+					: (this.visualizers[visualObject.attributes.type] ?? visualizeUnknownSharedObject);
 
 			const visualizerNode = new VisualizerNode(
-				sharedObject,
-				visualizationFunction,
+				// Double-casting `sharedObject` is necessary for `DataObject` visualization, because the `root` property is inaccessbile in `DataObject` (private).
+				isDataObject(visualObject)
+					? (visualObject as unknown as { readonly root: ISharedDirectory }).root
+					: (visualObject as ISharedObject),
```

Comment:
@Josmithr 

I couldn't assign an appropriate `visualObject` for an object that is a `PureDataObject` but not a `DataObject`, thus left `as ISharedObject` casting. 

I believe this will throw an runtime error if someone actually registers object type of above (e.g., `TreeDataObject`).

Reply:
We'll want to update `visualizeUnknownSharedObject` to accept the new allowed types then. This code shouldn't crash on unknown data.

Reply:
Correct me if I'm wrong here 😅, but my concern had more to do with the `VisualizerNode`. If we have to refactor the code, shouldn't we instead change the constructor type of the `VisualizerNode`? 

- `VisualizerNode` constructor requires an `ISharedObject`
- For `DataObject`, we can get an `ISharedObject` by accessing its root (which is `ISharedDirectory`, a type of `ISharedObject`)
- But for other `PureDataObjects` (like `TreeDataObject`), we don't know what property would give us an `ISharedObject` equivalent

Reply:
Closing this conversation after a separate conversation 😄 .